File: binary-search/1898-Maximum-Number-of-Removable-Characters.go
import "fmt"
func maximumRemovals(s string, p string, removable []int) int {
    left, right := 0, len(removable)
    ans := 0
    for left <= right {
        mid := (left + right) / 2
        if isValidRemoval(mid, s, p, removable) {
            ans = max(mid, ans)
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    return ans
}

func isValidRemoval(subsequence int, s string, p string, removable []int) bool {
    removal := make(map[int]bool)
    
    for _, r := range(removable[:subsequence]) {
        removal[r] = true
    }

    pPointer := 0
    for i := range s {
        if removal[i] {
            continue
        }
        if s[i] == p[pPointer] {
            pPointer += 1
            if pPointer == len(p) {
                return true
            }
        }
    }
    return false
}

File: binary-search/1898-Maximum-Number-of-Removable-Characters.py
class Solution:
    def maximumRemovals(self, s: str, p: str, removable: List[int]) -> int:
        left, right = 0, len(removable)
        ans = 0
        while left <= right:
            mid = (left + right) // 2
            if self.isValidRemoval(mid, s, p, removable):
                ans = max(mid, ans)
                left = mid + 1
            else:
                right = mid - 1
        return ans

    def isValidRemoval(self, subsequence: int, s: str, p: str, removable: List[int]) -> bool:
        removal = set(removable[:subsequence])
        p_pointer = 0
        for i in range(len(s)):
            if i in removal:
                continue
            if s[i] == p[p_pointer]:
                p_pointer += 1
                if p_pointer == len(p):
                    return True
        return False

